### ISSUE TITLE
Add rule to check number of columns in set operations

### ISSUE THREAD

**Issue by erevear** [sqlfluff/sqlfluff#1741]
Add rule to check that both queries in a set operation contain the same number of columns.

For example:

```
SELECT a, b, c
FROM test_table

UNION ALL

SELECT a, b
FROM test_table_2
```

The above produces a compilation error in Snowflake (`"invalid number of result columns for set operator input branches, expected 3, got 2 in branch 1"`), however it passes linting and parsing. 

Seems similar to rule L044, and could raise a parsing error.

**Comment by barrywhart** [comment]
I like the idea. I suggest that it should raise a linting error, not a parsing error. For technical reasons, rules don't raise parse errors. Parse errors are generally raised only by the parser, when the SQL does not obey the grammar (i.e. the dialect).

**Comment by jpy-git** [comment]
Good idea!
A clarification would be if one of the statements contains a wildcard we can't determine the number of columns, so let's not raise the lint error in that case as it would already be covered by L044.
